src/Adapters/Etw.Net451/EtwTelemetryModule.cs

```diff
@@ -0,0 +1,241 @@
+        /// </summary>
+        /// <param name="configuration">Module configuration.</param>
+        public void Initialize(TelemetryConfiguration configuration)
+        {
+            string errorMessage;
+            if (configuration == null)
+            {
+                errorMessage = string.Format(CultureInfo.InvariantCulture, "Argument {0} is required. The initialization is terminated.", nameof(configuration));
+                EventSourceListenerEventSource.Log.ModuleInitializationFailed(
+                    nameof(EtwTelemetryModule),
+                    errorMessage);
+                return;
+            }
+
+            if (this.isDisposed)
+            {
+                errorMessage = "Can't initialize a module that is disposed. The initialization is terminated.";
+                EventSourceListenerEventSource.Log.ModuleInitializationFailed(
+                    nameof(EtwTelemetryModule),
+                    errorMessage);
+                return;
+            }
+
+            bool? isProcessElevated = this.traceEventSession.IsElevated();
+            if (!isProcessElevated.HasValue || !isProcessElevated.Value)
+            {
+                errorMessage = "The process is required to be elevated to enable ETW providers. The initialization is terminated.";
+                EventSourceListenerEventSource.Log.RequiresToRunUnderPriviledgedAccount(nameof(EtwTelemetryModule));
+
+                // Throws so that user will be able to see the exception message in Output Window for debugging.
+                throw new UnauthorizedAccessException(errorMessage);
+            }
+
+            lock (this.lockObject)
+            {
+                this.client = new TelemetryClient(configuration);
+
+                // sdkVersionIdentifier will be used in telemtry entry as a identifier for the sender.
+                // The value will look like: etw:x.x.x-x
+                const string SdkVersionIdentifier = "etw:";
+                this.client.Context.GetInternalContext().SdkVersion = SdkVersionUtils.GetSdkVersion(SdkVersionIdentifier);
+
+                if (this.isInitialized)
+                {
+                    this.isInitialized = false;
+                    this.DisableProviders();
+                    this.enabledProviderIds.Clear();
+                    this.enabledProviderNames.Clear();
+                }
+
+                if (this.Sources.Count == 0)
+                {
+                    EventSourceListenerEventSource.Log.NoSourcesConfigured(moduleName: nameof(EtwTelemetryModule));
+                    return;
+                }
+
+                this.EnableProviders();
+                try
+                {
+                    // Start the trace session
+                    Task.Factory.StartNew(() => this.StartTraceEventSession(this.traceEventSession, this.client), TaskCreationOptions.LongRunning);
```

Comment:
@Dmitry-Matveev , good good question. I intention is to keep 1 task for the session for one instance, no matter how many times initialize is called. The implementation, though, will cause multiple tasks being started.

I think there are 2 ways to fix the issue. 1 is to close the task when initialize is called using a cancellation token; 2 is to keep the session but to enable / disable different providers on the session according the change of the configuration.

I think it the best performance + as much events being captured as possible to not interrupt the session when configure changes.

So, I'll update the code to start the session only when isInitialized flag is false.